src/AzureAppConfigurationImpl.ts

```diff
@@ -157,22 +174,51 @@ export class AzureAppConfigurationImpl extends Map<string, any> implements Azure
             keyValues.push([key, value]);
         }
 
-        this.clear(); // clear existing key-values in case of configuration setting deletion
+        this.#configMap.clear(); // clear existing key-values in case of configuration setting deletion
         for (const [k, v] of keyValues) {
-            this.set(k, v);
+            this.#configMap.set(k, v);
         }
+        // also construct hierarchical data object from map
+        this.#constructDataFromMap();
     }
 
     /**
      * Load the configuration store for the first time.
      */
     async load() {
         await this.#loadSelectedAndWatchedKeyValues();
-
         // Mark all settings have loaded at startup.
         this.#isInitialLoadCompleted = true;
     }
 
+    /**
+     * Construct hierarchical data object from map.
+     *
+     * @remarks
+     * This method is used to construct hierarchical data object from map, and it is called after the initial load and each refresh.
+     */
+    #constructDataFromMap() {
+        // clean data object in-place

```

Comment:
Do we want to do this in place? I assumed we were going to Object.Freeze data, and then fully rebuild it when updating (to make it immutable like the map).

Reply:
The purpose of in-place operation was to keep the appConfig.data reference unchanged, thus downstream consumer (e.g. feature manager) always get the updated value after a refresh, via appConfig.data.someKey.

Now after discussion, we don't expose a `data` property directly, instead we provide a helper function `constructData(options)` returning a data object as [mentioned](https://github.com/Azure/AppConfiguration-JavaScriptProvider/pull/49#issuecomment-1931196697). After a refresh, customers should update the reference of data object by themselves, with a proper onRefresh callback.
E.g.
```ts
// before adopting appconfig
config = JSON.parse(fs.readFileSync("config.json"))

console.log(config.server.port)

// after adopting appconfig
appConfig = await load(...)
config = appConfig.constructData()
appConfig.onRefresh(() => { config = appConfig.constructData() }) // the callback ensure the config is always up-to-date

console.log(config.server.port)
```

Reply:
I don't think this change solves much- except maybe helping guide user expectations. 

For the user, there is functionally no difference between:
```javascript
config = appConfig.constructData();
// and
config = appConfig.data;
```

The separator and prefix are a bonus, but they could be placed in the load function options (and probably should anyway so you don't repeat it everytime you call "constructData")
```javascript
appConfig = await load({
  "connection_string": "~~~",
  "separator": ":",
})
```
And this aligns with python's `load` funciton.

Reply:
It’s unlike python. In JS here the preferred way is to use get(key). In such use case, the key is full key matching app config store, and the customer doesn’t need to care about the separator when calling load(). The separator is only related to how we restore the hierarchy in the data object if we need one.

Moreover, about constructing data object, as I mentioned there might be edge case that ambiguity occurs and you have to drop some information or directly fail. It’s unreasonable to fail the load() as the key values are valid in app config store and the get(key) works as expected. With a function constructData(…), we can allow user to explicitly specify the behavior here and fail this function instead of the load(). It’s more reasonable and clear.

The prefix is for “configuration section” support. In node-config, there is a popular usage that we put configs of different components into an all-in-one config.json file, with component names as top level keys. Usage is like dbConfig=config.get<DbConfig>(“db”), where DbConfig defines the shape of the section. Then in app code, dbConfig.url, dbConfig.username etc are used. The prefix option here offers to only construct part of the loaded key values into a data object that customers do care about. And it’s also dedicated to the data object construction instead of the load().